server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,590 @@
+    return 0
+
+
+# Archive Hierarchy
+def verify_subdirs(hier, controller, directories):
+    linkdirs = sorted(hier.config.LINKDIRS.split(" "))
+    if directories:
+        for dirent in directories:
+            if all(
+                [
+                    Path(hier.path, controller, dirent).exists(),
+                    dirent != "_QUARANTINED",
+                    not dirent.startswith("WONT-INDEX"),
+                ]
+            ):
+                if dirent not in linkdirs:
+                    hier.add_unexpected_controllers(
+                        Hierarchy.UNEXPECTED_DIRS, controller, dirent
+                    )
+    else:
+        hier.add_unexpected_controllers(
+            Hierarchy.SUBDIR_STATUS_INDICATORS, controller, "subdirs"
+        )
+
+    return 0
+
+
+def verify_prefixes(hier, controller):
+    prefix_dir = Path(hier.path, controller, ".prefix")
+    if not prefix_dir.exists():
+        return 0
+    if not prefix_dir.is_dir():
+        hier.add_unexpected_controllers(
+            Hierarchy.PREFIX_STATUS_INDICATORS, controller, "prefix_dir"
+        )
+        return 1
+
+    prefixes = prefix_dir.glob("*")
+
+    for prefix_path in prefixes:
+        if not prefix_path.name.startswith("prefix.") and not prefix_path.name.endswith(
+            ".prefix"
+        ):
+            hier.add_unexpected_controllers(
+                Hierarchy.NON_PREFIXES, controller, prefix_path.name
+            )
+        elif prefix_path.name.startswith("prefix."):
+            hier.add_unexpected_controllers(
+                Hierarchy.WRONG_PREFIXES, controller, prefix_path.name
+            )
+
+    return 0
+
+
+# 'hier' is an object of ArchiveHierarchy class imported from hierarchy.py
+def verify_archive(hier):
+
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    # now check each "good" controller and get the tarballs it contains
+    for controller in hier.controllers:
+        direct_entries = glob.iglob(os.path.join(hier.path, controller, "*"))
+        hidden_entries = Path(hier.path, controller).glob(".*")
+        if hidden_entries:
+            for hid_entries_path in hidden_entries:
+                if hid_entries_path.is_file():
+                    hier.add_unexpected_controllers(
+                        Hierarchy.UNEXPECTED_OBJECTS, controller, hid_entries_path.name,
+                    )
+        controller_subdir = list()
+        for item in direct_entries:
+            item_path = Path(item)
+            if item_path.is_dir():
+                controller_subdir.append(item_path.name)
+            elif item_path.is_symlink():
+                try:
+                    item_p = item_path.resolve(strict=True)
+                except FileNotFoundError:
+                    print(
+                        f"{item_path.name}_path, '{item_path}', does not resolve to a real location"
+                    )
+                    return 1
```

Comment:
We don;t want to return here: we want to find all of the problems and report them. So either log this error or (better IMO) add another category of errors to the hierarchy and collect them there so that they can be reported later.

Reply:
So, should I create another dict value "error" and whatever error will occur I should record that in that?